### PR TITLE
Adds Score Details

### PR DESCRIPTION
**Pull request by victorhuangwq**
## Description

This pull request enhances the submissions management interface by adding a modal to display detailed score information for each student's submission. It includes the necessary JavaScript, CSS, HTML, and backend changes to fetch and display the data, as well as modifications to the routes file. The modal displays version number, submission date, final score, problem scores, late penalty, tweak, and action buttons for each submission.
image

Currently, it's visually not to specification yet, and will require further modifications.

## Motivation and Context

Part of lynn's efforts to redesign manage submissions. 

## How Has This Been Tested?

1. Verified that the detailed score information modal was triggered when clicking the "zoom_in" icon in the submissions table.
2. Checked the displayed data in the modal for accuracy and consistency with the actual submission data.
3. Tested the sorting and filtering functionalities within the modal's DataTable, ensuring they worked as expected.
4. Verified that the modal properly displayed autograded and non-autograded problems.
5. Checking that the tweaks are clickable, and properly displayed after editing them
6. Tried the view source and download button and it works.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have run rubocop for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://github.com/autolab/docs)
- [ ] I have updated the documentation accordingly, included in this PR

## Other issues / help required

If unsure, feel free to submit first and we'll help you along.

### CODE REVIEW

File: app/assets/javascripts/manage_submissions.js [hunk 1]
```diff
@@ -1,4 +1,171 @@
...
+
+      const submissions_body = data.submissions.map((submission) => {
+        
+        let tweak_value = data?.tweaks[submission.id]?.value ?? "None";
+        if (tweak_value != "None" && tweak_value > 0) {
+          tweak_value = `+${tweak_value}`;
+        }
+
+        // Convert to human readable date with timezone 
+        const human_readable_created_at = moment(submission.created_at).format('MMM Do YYYY, h:mm:ss a z Z');
```
Comment (lykimchee) [inline app/assets/javascripts/manage_submissions.js:68]: Is this the finalized date format? I remember in the team meeting we discussed a specific format but don't remember what the final format we decided on was.

**Review by lykimchee — CHANGES_REQUESTED** [review]
For excused students, the score popup looks like this:
image

After modifying the assessment (added a problem), pressing on score popup gave me this:
image

Also, can there be an X icon to close popup? Figma design screenshot below:
image

Also, list of styling issues (I know some of the styling is going to be done later, but if any of these, particularly the ones related to adding divs/spans/HTML structure, could be resolved that would be helpful!)
- problem max score style should be different (also, Autograded ends up being in italics)
- problems on the table should have different color backgrounds, Figma design screenshot below:

image

- table bottom spacing with the popup
- table data is slightly misaligned upon adding proper submissions-th submissions-td class

image

- table header text formatting/style
- add pencil icons next to each problem score if not auto graded, Figma design screenshot below:

image

File: app/assets/javascripts/manage_submissions.js [hunk 2]
```diff
@@ -1,4 +1,171 @@
...
+    $('#score-details-modal').modal('open');
+
+    // Fetch data and render it in the modal 
+    get_score_details(course_user_datum_id).then((data) => {
+      console.log(data);
+
+      const problem_headers = data.submissions[0].problems.map((problem) => {
+        const max_score = problem.max_score;
+        const autograded = problem.grader_id < 0 ? " (Autograded)" : "";
+        return `<th class="submission-th">
```
[the remark this answers is not in the document]
Comment (victorhuangwq) [reply app/assets/javascripts/manage_submissions.js:53]: nice catch

File: app/assets/javascripts/manage_submissions.js (hunk 1 above)
Comment (victorhuangwq) [reply app/assets/javascripts/manage_submissions.js:68]: I updated it to match

File: app/assets/javascripts/manage_submissions.js [hunk 3]
```diff
@@ -1,4 +1,171 @@
...
+              </td>
+              <td class="submission-td">
+                ${view_button}
+                ${download_button}
+              </td>
+            </tr>`;
+      }).join('');
+
+      const submissions_table =
+        ` <p>Click on non-autograded problem scores to edit or leave a comment </p>
```
[the remark this answers is not in the document]
Comment (victorhuangwq) [reply app/assets/javascripts/manage_submissions.js:131]: done

File: config/routes.rb [hunk 4]
```diff
@@ -150,6 +150,7 @@
         collection do
           get "downloadAll"
           get "missing"
+          get "score_details"
```
[the remark this answers is not in the document]
Comment (victorhuangwq) [reply config/routes.rb:153]: there's sadly no standardization on whether we use camelcase or underscores in Autolab... and it's too late to change that, for now.

File: app/assets/javascripts/manage_submissions.js [hunk 5]
```diff
@@ -1,4 +1,171 @@
...
+  return new Promise((resolve, reject) => {
+    $.ajax({
+      url: manage_submissions_endpoints['score_details'],
+      type: 'GET',
+      data: { cuid: course_user_datum_id },
+      success: function (data) {
+        resolve(data);
+      },
+      error: function (err) {
+        console.log(err);
```
[the remark this answers is not in the document]
Comment (victorhuangwq) [reply app/assets/javascripts/manage_submissions.js:15]: let me remove them

**Comment by victorhuangwq** [comment]
> For excused students, the score popup looks like this:
I couldn't quite replicate this. what i did:
- I excused a student from grade submissions
- i checked it in manage submissions, and there isn't any errors

> After modifying the assessment (added a problem), pressing on score popup gave me this:
Interesting case, I fixed it. Basically, the issue was there there isn't a score assigned to a submission when a new problem is added, so it errors out. I now make it return undefined
image

**Comment by victorhuangwq** [comment]
> Also, can there be an X icon to close popup? Figma design screenshot below:
Added a cross button, but it's not exactly styled still

**Comment by victorhuangwq** [comment]
> problem max score style should be different (also, Autograded ends up being in italics)
I made the second row italics. But the bolding is a little more involved I think

**Comment by victorhuangwq** [comment]
> table bottom spacing with the popup
not quite sure what you are referring to here

**Comment by victorhuangwq** [comment]
> table data is slightly misaligned upon adding proper submissions-th submissions-td class
added padding

**Comment by victorhuangwq** [comment]
> add pencil icons next to each problem score if not auto graded, Figma design screenshot below:
I think this would go better in the annotation popup PR

**Review by lykimchee — APPROVED** [review]
Sorry for the delay, I was out for a while! LGTM— I can take care of any styling stuff later, tysm Victor!! :'))

**Pull request merged**